src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -397,23 +397,21 @@ public void ExecuteSuppressionAction(DiagnosticSuppressor suppressor, ImmutableA
         {
             Debug.Assert(compilationEvent is CompilationStartedEvent || compilationEvent is CompilationCompletedEvent);
 
-            AnalyzerStateData? analyzerState = null;
-
-            try
+            if (TryStartProcessingEvent(compilationEvent, analyzer, analysisScope, analysisState, out var analyzerState))
```

Comment:
@AlekseyTs State machine transitions should now be very clear:
1. We attempt to `TryStartProcessingEvent` to get hold of analyzer state, with `StateKind.InProcess`.
2. If we succeed, then we attempt to execute relevant analyzer callbacks for the compilation event
3. After executing callbacks, we transition the state to `StateKind.ReadyToProcess` state
4. Finally, we attempt to mark completion of the event for the analyzer and try to transition to `StateKind.FullyProcessed` and return true only if we succeed.
5. If we fail at Step 1, which happens when either another thread is executing steps 2-4 OR analyzer state has already completed and transitioned to `FullyProcessed`, we don't attempt to execute analysis but instead return `IsEventComplete` to handle the latter case.